nifi-nar-bundles/nifi-extension-utils/nifi-record-utils/nifi-standard-record-utils/src/main/java/org/apache/nifi/serialization/SplitRecordSetHandler.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.serialization;
+
+import org.apache.nifi.serialization.record.Record;
+import org.apache.nifi.serialization.record.RecordSet;
+
+import java.io.IOException;
+
+public abstract class SplitRecordSetHandler {
+    final int maximumChunkSize;
+
+    protected SplitRecordSetHandler(int maximumChunkSize) {
+        if (maximumChunkSize < 1) {
+            throw new IllegalArgumentException("The maximum chunk size must be a positive number");
+        }
+
+        this.maximumChunkSize = maximumChunkSize;
+    }
+
+    public final RecordHandlerResult handle(final RecordSet recordSet) throws Exception {
+        return handle(recordSet, 0);
+    }
+
+    public final RecordHandlerResult handle(final RecordSet recordSet, final int alreadyProcessedChunks) throws IOException {
+        Record record;
+        int currentChunkNumber = 1;
+        int currentChunkSize = 0;
+
+        while ((record = recordSet.next()) != null) {
+            if (currentChunkSize < maximumChunkSize) {
+                addToChunk(record);
+                currentChunkSize++;
+            } else {
+                try {
+                    handleChunk(alreadyProcessedChunks >= currentChunkNumber);
+                    addToChunk(record);
+                    currentChunkNumber++;
+                    currentChunkSize = 1;
+                } catch (final SplitRecordSetHandlerException e) {
+                    return new RecordHandlerResult(currentChunkNumber - 1, e);
+                }
+            }
+        }
+
+        // Handling the last, not fully filled chunk
+        if (currentChunkSize != 0) {
+            try {
+                handleChunk(alreadyProcessedChunks >= currentChunkNumber);
+            } catch (final SplitRecordSetHandlerException e) {
+                return new RecordHandlerResult(currentChunkNumber - 1, e);
+            }
+        }
+
+        // The incoming RecordSet was empty
+        if (currentChunkSize == 0 && currentChunkNumber == 1) {
+            currentChunkNumber = 0;
+        }
+
+        return new RecordHandlerResult(currentChunkNumber);
```

Comment:
In my opinion, it would be more straightforward this way (copying the whole method body because there are multiple small changes):
```
Record record;
int currentChunkNumber = 0;
int currentChunkSize = 0;

while ((record = recordSet.next()) != null) {
    addToChunk(record);
    currentChunkSize++;

    if (currentChunkSize == maximumChunkSize) {
        try {
            handleChunk(alreadyProcessedChunks > currentChunkNumber);
            currentChunkNumber++;
            currentChunkSize = 0;
        } catch (final SplitRecordSetHandlerException e) {
            return new RecordHandlerResult(currentChunkNumber, e);
        }
    }
}

// Handling the last, not fully filled chunk
if (currentChunkSize != 0) {
    try {
        handleChunk(alreadyProcessedChunks > currentChunkNumber);
        currentChunkNumber++;
    } catch (final SplitRecordSetHandlerException e) {
        return new RecordHandlerResult(currentChunkNumber, e);
    }
}

return new RecordHandlerResult(currentChunkNumber);
```
Though I do not insist on it at all.
If you keep the original one, please fix the comment:
`// Handling the last, not fully filled chunk` => `// Handling the last chunk`
because it may or may not be fully filled in that implementation.